File: tests/interfile/kotlin/test_hof_comprehensive_kotlin_split/zz_main.kt
// Comprehensive HOF test for Kotlin: Custom and built-in higher-order functions
// All of these should detect taint flow from source() to sink()

// ===== Custom HOF Functions =====

fun <T> directCall(callback: (T) -> Unit, value: T) {
    callback(value)
}

// ===== Test Cases =====


// ===== Built-in collection functions =====




// ===== Complex Example =====



// Stub functions
fun source(): String = "tainted"

// ===== Top-level HOF Tests =====
// These test HOF callback detection at top level (outside any function)

// Top-level lambda callback
// ruleid: test-hof-taint
val toplevelSink = { x: String -> sink(x) }
val toplevelResult1 = toplevelSink(source())

// Top-level method HOF (forEach with named callback)

val toplevelItems = listOf(source())
val toplevelResult2 = toplevelItems.forEach(::toplevelHandler)

// Top-level user-defined HOF
val toplevelResult3 = directCall(::toplevelHandler, source())

File: tests/interfile/julia/test_hof_comprehensive_julia_split/zz_main.jl
# Comprehensive HOF test for Julia: Custom and built-in higher-order functions
# All of these should detect taint flow from source() to sink()

# ===== Custom HOF Functions =====

# Manual loop implementation

# Delegates to built-in (tests ToSinkInCall propagation)



# ===== Test Cases =====

# Named function for testing

# Test custom HOF with manual loop + lambda

# Test custom HOF with manual loop + named function

# Test custom HOF delegating to built-in + lambda

# Test custom HOF delegating to built-in + named function



# ===== Built-in functions =====




# ===== Complex Example =====



# Stub functions


# ===== Top-level HOF Tests =====
# These test HOF callback detection at module level (outside any function)

# Top-level lambda callback
# ruleid: test-hof-taint
toplevel_sink = x -> sink(x)
toplevel_sink(source())

# Top-level function HOF (map with named callback)

toplevel_items = [source()]
map(toplevel_handler, toplevel_items)

# Top-level user-defined HOF
customForEach(toplevel_items, toplevel_handler)
